src/lib/update-lock.ts

```diff
@@ -96,69 +87,93 @@ function dispose(
-						const lockDir = getPathOnHost(lockPath(appId));
-						return Bluebird.resolve(fs.readdir(lockDir))
-							.catchReturn(ENOENT, [])
-							.mapSeries((serviceName) => {
-								return Bluebird.mapSeries(
-									lockFilesOnHost(appId, serviceName),
-									(tmpLockName) => {
-										return (
-											Bluebird.try(() => {
-												if (force || lockOverride) {
-													return lockfile.unlock(tmpLockName);
-												}
-											})
-												.then(() => {
-													return lockfile.lock(tmpLockName, LOCKFILE_UID);
-												})
-												// If lockfile exists, throw a user-friendly error.
-												// Otherwise throw the error as-is.
-												// This will interrupt the call to Bluebird.using, so
-												// dispose needs to be called even though it's referenced
-												// by .disposer later.
-												.catch((error) => {
-													return dispose(release, appId).throw(
-														lockfile.LockfileExistsError.is(error)
-															? new UpdatesLockedError(
-																	`Lockfile exists for ${JSON.stringify({
-																		serviceName,
-																		appId,
-																	})}`,
-															  )
-															: (error as Error),
-													);
-												})
-										);
-									},
-								);
-							});
-					})
-					.disposer((release: () => void) => dispose(release, appId));
-			})
-			.catch((err) => {
-				throw new InternalInconsistencyError(
-					`Error getting lockOverride config value: ${err?.message ?? err}`,
-				);
-			});
-	};
+	// Sort appIds so they are always locked in the same sequence
+	const sortedIds = appIdsToLock.sort();
 
-	const disposer = takeTheLock();
+	let lockOverride: boolean;
+	try {
+		lockOverride = await config.get('lockOverride');
+	} catch (err) {
+		throw new InternalInconsistencyError(
+			`Error getting lockOverride config value: ${err?.message ?? err}`,
+		);
+	}
 
-	return Bluebird.using(disposer, fn as () => PromiseLike<T>);
+	const releases = new Map<number, () => void>();
+	try {
+		for (const id of sortedIds) {
+			const lockDir = getPathOnHost(lockPath(id));
+			// Acquire write lock for appId
+			releases.set(id, await writeLock(id));
+			// Get list of service folders in lock directory
+			let serviceFolders: string[] = [];
+			try {
+				serviceFolders = await fs.readdir(lockDir);
+			} catch (err) {
+				if (ENOENT(err)) {
+					// Default to empty list of folders
+					serviceFolders = [];
+				} else {
+					throw err;
+				}
+			}
+			// Attempt to create a lock for each service
+			for (const service of serviceFolders) {
+				await lockService(id, service, { force, lockOverride });
+			}
+		}
+		// Resolve the function passed
+		return Promise.resolve(fn());
+	} finally {
+		// Cleanup locks
+		for (const [id, release] of releases.entries()) {
+			// Release lock on application
+			release();
+			// Dispose of locks in each service within the application
+			await dispose(id);
```

Comment:
> release was done within a finally before.

Release removes application locks
Dispose removes service locks

The previous code passed release to dispose so that it (release) would be called after disposing (release services). It wasn't mentioned but this PR changes release to be called before disposing so I'll change that to be more like the old behaviour.

Regarding error handling though, this code is the same I believe. 

Previously, when locking a service, if it throws an error it was [caught and logged](https://github.com/balena-os/balena-supervisor/blob/33ad877c67a639024a35cd4aa8452d280f757a45/src/lib/update-lock.ts#L136-L145). The current behaviour catches the error when locking a service too but with the exception of not releasing the application yet. This is because the finally block always runs and that calls release. The old code would release twice in fact. That never threw an error because of [this](https://github.com/71104/rwlock/blob/master/src/lock.js#L180) check in the rwlock module.

So to clarify, the old code would have locks left in memory/disk since it just logs when there was a problem releasing service locks.

Reply:
> That is the reason the call to release was done within a finally before.

Unless there's another behaviour of this but it's hard to read since the old code didn't use finally it used dispose.

Reply:
right, basically: 

old code: might not clear all the service locks but should always clear application lock
new code: might not clear all service locks AND might not clear application lock if service lock clearing has a problem

Reply:
resolving because we discussed in another chat. I passed release to dispose which uses finally to always call the release if any unlocking of a service fails. This is the exact same behavior (and code) as before.